v2/ghutils/github.go

```diff
@@ -0,0 +1,197 @@
+	// ref: https://golang.org/ref/mod#incompatible-versions
+	args.Version = strings.TrimRight(args.Version, "+incompatible")
+	url := fmt.Sprintf(
+		template,
+		repo.Owner,
+		repo.Name,
+		parseGoModulePseudoVersion(args.Version),
+		args.Path)
+	if args.LineStart > 0 {
+		if args.Raw {
+			return "", fmt.Errorf("GitHubRepo.RemoteUrl(%+v): LineStart and LineEnd not supported for url to raw content.", args)
+		}
+		url = url + fmt.Sprintf("#L%v-L%v", args.LineStart, args.LineEnd)
+	}
+	return url, nil
+}
+
+// Reference: https://golang.org/ref/mod#pseudo-versions
+// vX.0.0-yyyymmddhhmmss-abcdefabcdef is used when there is no known base version. As with all versions, the major version X must match the module's major version suffix.
+// vX.Y.Z-pre.0.yyyymmddhhmmss-abcdefabcdef is used when the base version is a pre-release version like vX.Y.Z-pre.
+// vX.Y.(Z+1)-0.yyyymmddhhmmss-abcdefabcdef is used when the base version is a release version like vX.Y.Z. For example, if the base version is v1.2.3, a pseudo-version might be v1.2.4-0.20191109021931-daa7c04131f5.
+var psuedoVersionPattern = regexp.MustCompile(`^v[0-9]+\.[0-9]+\.[0-9]+-.*[0-9]{14}-(?P<commit>[a-f0-9]{11,12})$`)
+
+func parseGoModulePseudoVersion(version string) string {
+	if version == "" {
+		return "main" // default to content in main branch
+	}
+	// Parse version like v0.0.0-20210108172934-df6aa8a2788b to commit hash:
+	// df6aa8a2788b.
+	matches := psuedoVersionPattern.FindStringSubmatch(version)
+	if len(matches) == 2 {
+		// matches[0] is regex match, matches[1] is result of the capture group.
+		return matches[1]
+	}
+	return version
+}
+
+var githubUrlPattern = regexp.MustCompile(`^(https://)?(www\.)?github.com/(?P<repo>[^/]+/[^/]+)/blob/(?P<path>[^#]*)(?P<hash>#.*)?$`)
+var githubLinePattern = regexp.MustCompile(`^#L(?P<linestart>[0-9]+)-L(?P<lineend>[0-9]+)$`)
+
+func GithubDownloadUrl(url string) (downloadUrl string, lineStart int, lineEnd int, err error) {
+	matches := githubUrlPattern.FindStringSubmatch(url)
+	if len(matches) > 0 {
+		repo := matches[3]
+		path := matches[4]
+		hash := matches[5]
+		if hash == "" {
+			return fmt.Sprintf("https://github.com/%s/raw/%s", repo, path), 0, 0, nil
+		}
+		lineMatches := githubLinePattern.FindStringSubmatch(hash)
+		if len(lineMatches) == 0 {
+			return "", 0, 0, fmt.Errorf("getGithubDownloadUrl(%q): cannot find line numbers in hash", url)
+		}
+		// line start and line end included
+		lineStart, err := strconv.ParseInt(lineMatches[1], 10, 0)
+		if err != nil {
+			return "", 0, 0, err
+		}
+		lineEnd, err := strconv.ParseInt(lineMatches[2], 10, 0)
+		if err != nil {
+			return "", 0, 0, err
+		}
+		return fmt.Sprintf("https://github.com/%s/raw/%s", repo, path), int(lineStart), int(lineEnd), nil
+	}
+	return "", 0, 0, nil
+}
+
+// TODO: this downloads url content in memory.
+// We might need optimization in the future.
+func SmartDownload(url string) (string, error) {
```

Comment:
P1
Completely agree. I didn't do this initially was because

1. I didn't include enough information in the csv to find the local license file.
2. I liked that the download phase verifies all urls are valid.

This topic needs some proper design, so putting as P1.